_posts/2020-08-01-data-science-meets-devops.md

```diff
@@ -0,0 +1,755 @@
+
+
+## Further Details
+
+
+### Background
+
+Before diving into the details of our solution we want to provide some context around:
+
+
+
+1. What reconcilers are and why we use them
+2. What GitOps is and why we adopted this pattern
+
+
+### Building Resilient Systems With Reconcilers
+
+A reconciler is a control pattern that has proven to be immensely useful for building resilient systems. The reconcile pattern is [at the heart of how Kubernetes works](https://book.kubebuilder.io/cronjob-tutorial/controller-overview.html). Figure 3 illustrates how a reconciler works. A reconciler works by first observing the state of the world; e.g. what model is currently deployed. The reconciler then compares this against the desired state of the world and computes the diff; e.g the model with label “version=20200724” should be deployed, but the model currently deployed has label “version=20200700”. The reconciler then takes the action necessary to drive the world to the desired state; e.g. open a pull request to change the deployed model.
+
+
+<table>
+  <tr>
+   <td>
+<img src="/images/2020-08-01-data-science-meets-devops/fig3.reconciler.png" width="" alt="alt_text" title="image_tooltip">
+
+   </td>
+  </tr>
+  <tr>
+   <td>Figure 3. Illustration of the reconciler pattern as applied by our deployer.
+   </td>
+  </tr>
+</table>
+
+
+Reconcilers have proven immensely useful for building resilient systems because a well implemented reconciler provides a high degree of confidence that no matter how a system is perturbed it will eventually return to the desired state.
+
+
+### There Are No DAGs
+
+The declarative nature of controllers means data can flow through a series of controllers  without needing to explicitly create a DAG. In lieu of a DAG, a series of data processing steps can instead be expressed as a set of desired states, as illustrated in Figure 4 below:
+
+
+<table>
+  <tr>
+   <td>
+
+<img src="/images/2020-08-01-data-science-meets-devops/fig4.data-pipeline.png
+" width="" alt="alt_text" title="image_tooltip">
+
+   </td>
+  </tr>
+  <tr>
+   <td>Figure 4 illustrates how pipelines can emerge from independent controllers without explicitly encoding a DAG. Here we have two completely independent controllers. The first controller ensures that for every element a<sub>i</sub> there should be an element b<sub>i</sub>. The second controller ensures that for every element b<sub>i</sub> there should be an element c<sub>i</sub>. 
+   </td>
+  </tr>
+</table>
+
+
+This reconciler-based paradigm offers the following benefits over many traditional DAG-based workflows:
+
+
+
+*   **Resilience against failures**:  the system continuously seeks to achieve and maintain the desired state.  This state refers to both infrastructure (via Kubernetes manifests) and computation to be performed.  This is important in machine learning workflows where infrastructure is often strongly coupled with code (i.e. GPUs, databases, etc).
```

Comment:
@hamelsmu 

> This is important in machine learning workflows where infrastructure is often strongly coupled with code (i.e. GPUs, databases, etc).

I didn't quite understand what you were trying to say here. Could you elaborate please?

Reply:
@jlewi  Maybe it is not worded well.  I was trying to highlight that the system is self-healing with regards to infrastructure b/c of Kubernetes (spinning up pods, etc. )  and also resiliency in the form of achieving the desired state with respect to the jobs that need to be run.   

I then was trying to make the point that infrastructure and code are two common failure modes for machine learning so this is important.  

Feel free to delete/modify/etc. in a way that makes more sense

Reply:
we resolved this